subsys/net/lib/http/http_server.c

```diff
@@ -0,0 +1,1508 @@
+				   struct http_client_ctx *client)
+{
+	struct http_frame *frame = &client->current_frame;
+	struct http_stream_ctx *stream;
+
+	stream = find_http_stream_context(client, frame->stream_identifier);
+	if (!stream) {
+		LOG_DBG("|| stream ID ||  %d", frame->stream_identifier);
+
+		stream = allocate_http_stream_context(client, frame->stream_identifier);
+		if (!stream) {
+			LOG_DBG("No available stream slots. Connection closed.");
+			close_client_connection(server, client);
+
+			return -ENOMEM;
+		}
+	}
+
+	if (settings_end_headers_flag(client->current_frame.flags) &&
+	    settings_end_stream_flag(client->current_frame.flags)) {
+		client->server_state = HTTP_SERVER_FRAME_HEADERS_STATE;
+	} else {
+		client->server_state = HTTP_SERVER_FRAME_HEADER_STATE;
+	}
+
+	return 0;
+}
+
+int enter_http_frame_continuation_state(struct http_client_ctx *client)
+{
+	client->server_state = HTTP_SERVER_FRAME_CONTINUATION_STATE;
+
+	return 0;
+}
+
+int enter_http_frame_window_update_state(struct http_client_ctx *client)
+{
+	client->server_state = HTTP_SERVER_FRAME_WINDOW_UPDATE_STATE;
+
+	return 0;
+}
+
+int enter_http_frame_priority_state(struct http_client_ctx *client)
+{
+	client->server_state = HTTP_SERVER_FRAME_PRIORITY_STATE;
+
+	return 0;
+}
+
+int enter_http_frame_rst_stream_state(struct http_server_ctx *server,
+				      struct http_client_ctx *client)
+{
+	client->server_state = HTTP_SERVER_FRAME_RST_STREAM_STATE;
+
+	return 0;
+}
+
+int enter_http_frame_goaway_state(struct http_server_ctx *server,
+				  struct http_client_ctx *client)
+{
+	client->server_state = HTTP_SERVER_FRAME_GOAWAY_STATE;
+
+	return 0;
+}
+
+int enter_http_http_done_state(struct http_server_ctx *server,
+			       struct http_client_ctx *client)
+{
+	client->server_state = HTTP_SERVER_DONE_STATE;
+
+	return 0;
+}
+
+int handle_http_preface(struct http_client_ctx *client)
+{
+	LOG_DBG("HTTP_SERVER_PREFACE_STATE.");
+
+	if (client->offset < sizeof(preface) - 1) {
+		/* We don't have full preface yet, get more data. */
+		return -EAGAIN;
+	}
+
+	if (strncmp(client->buffer, preface, sizeof(preface) - 1) != 0) {
+		client->server_state = HTTP_SERVER_REQUEST_STATE;
+	} else {
+		client->server_state = HTTP_SERVER_FRAME_HEADER_STATE;
+		client->offset -= sizeof(preface) - 1;
+
+		memmove(client->buffer, client->buffer + sizeof(preface) - 1,
+			client->offset);
+	}
+
+	return 0;
+}
+
+struct http_resource_detail *get_resource_detail(const char *path,
+						 int *path_len)
+{
+	HTTP_SERVICE_FOREACH(service) {
+		HTTP_SERVICE_FOREACH_RESOURCE(service, resource) {
+			int len = strlen(resource->resource);
+
+			if (strncmp(path, resource->resource, len) == 0) {
+				*path_len = len;
+				return resource->detail;
+			}
+		}
+	}
+
+	return NULL;
+}
+
+int handle_http1_static_resource(struct http_resource_detail_static *static_detail,
+				 int client_fd)
+{
+#define RESPONSE_TEMPLATE			\
+	"HTTP/1.1 200 OK\r\n"			\
+	"Content-Type: text/html\r\n"		\
```

Comment:
The Content-Type needs to be configurable per resource. Browsers refuse to run javascripts sent with text/html content-type.